lib/usd/translators/meshWriter_BlendShapes.cpp

```diff
@@ -0,0 +1,931 @@
+                    CHECK_MSTATUS_AND_RETURN_IT(stat);
+                    const float* meshPts = fnMesh.getRawPoints(&stat);
+                    CHECK_MSTATUS_AND_RETURN_IT(stat);
+                    (void)meshPts;
+                }
+            }
+        }
+    }
+
+    // NOTE: (yliangsiew) Restore the weight changes that we were doing above.
+    for (unsigned int i = 0; i < indicesChanged.length(); ++i) {
+        stat = fnBS.setWeight(indicesChanged[i], origWeightVals[i]);
+        CHECK_MSTATUS_AND_RETURN_IT(stat);
+    }
+#endif
+
+    MDataHandle dhInputComponentsTarget = plgInputComponentsTarget.asMDataHandle(&stat);
+
+    CHECK_MSTATUS_AND_RETURN_IT(stat);
+    MObject inputComponentsTargetIndices = dhInputComponentsTarget.data();
+    if (inputComponentsTargetIndices.isNull()
+        || !inputComponentsTargetIndices.hasFn(MFn::kComponentListData)) {
+        return MStatus::kFailure;
+    }
+    MFnComponentListData fnComponentListData(inputComponentsTargetIndices, &stat);
+    CHECK_MSTATUS_AND_RETURN_IT(stat);
+    unsigned int numIndices = fnComponentListData.length();
+    if (numIndices == 0) {
+        return MStatus::kFailure;
+    }
+
+    targetOffsetIndices.clear();
+    MIntArray indices;
+    for (unsigned int i = 0; i < numIndices; ++i) {
+        MObject                   curComponent = fnComponentListData[i];
+        MFnSingleIndexedComponent fnSingleIndexedComponent(curComponent, &stat);
+        CHECK_MSTATUS_AND_RETURN_IT(stat);
+        stat = fnSingleIndexedComponent.getElements(indices);
+        CHECK_MSTATUS_AND_RETURN_IT(stat);
+        for (unsigned int j = 0; j < indices.length(); ++j) {
+            targetOffsetIndices.push_back(indices[j]);
+        }
+    }
+
+    MPlug plgInputPointsTarget = mayaFindChildPlugWithName(plgInputTargetItem, "inputPointsTarget");
+    MDataHandle dhInputPointsTarget = plgInputPointsTarget.asMDataHandle(&stat);
+    CHECK_MSTATUS_AND_RETURN_IT(stat);
+
+    MObject inputPointsTarget = dhInputPointsTarget.data();
+    assert(inputPointsTarget.hasFn(MFn::kPointArrayData));
+
+    MFnPointArrayData fnPtArrayData(inputPointsTarget, &stat);
+    CHECK_MSTATUS_AND_RETURN_IT(stat);
+
+    MPointArray  mptTargetOffsets = fnPtArrayData.array();
+    unsigned int numOffsets = mptTargetOffsets.length();
+    if (numOffsets == 0) { // NOTE: (yliangsiew) There cannot possibly be _no_ offsets in the
+                           // blendshape, since we have indices for the offsets.
+        return MStatus::kFailure;
+    }
+    targetOffsets.resize(numOffsets);
+    for (unsigned int i = 0; i < numOffsets; ++i) {
+        MPoint curPt = mptTargetOffsets[i];
+        targetOffsets[i] = GfVec3f(curPt.x, curPt.y, curPt.z);
+    }
+
+    MPlug plgTargetGeom = mayaFindChildPlugWithName(plgInputTargetItem, "inputGeomTarget");
+    if (plgTargetGeom.isNull()) {
+        return MStatus::kFailure;
+    }
+    MObject targetGeom = plgTargetGeom.asMObject();
```

Comment:
You're right, animated targets are another limitation I didn't think about here. After some internal discussion, it appears that UsdSkelBlendShape offsets and normalOffsets are currently uniforms, which means that they can't be time-sampled apparently and thus it would be tricky to support animated blendshape targets the way Maya does. I think the suggestion you gave about blocking an export of animated shape targets (or warning) makes the most sense for now and I'll see about adding this. What do you think? (Warning VS hard block of export)

Reply:
Warning vs. a blocking error should be an export configuration, where the default value is to fail but the user can override the value and get the best-effort export that can be produced.

The same approach can help with some of the other ambiguous bits of code around what happens when there are multiple upstream blend shapes, or if there is a wrap deformer between blend shapes. In those cases you can error out saying the exporter doesn't support that pattern, but you can try forcing it and see what you get.

Reply:
When you say "export configuration", do you mean having an additional flag for the mayaUSDExport command i.e. ignoreErrors or similar? I can certainly add a catch-all option for users to override what happens; or are you referring to something else?

Reply:
Yes an additional flag on the export command.

Reply:
Do you think it would make sense to have a single flag that all aspects of the exporter could follow? Something like a `--strict`. 

We could make blendshapes be the first aspect of the exporter that follows that flag, but it would potentially be nicer UX to have a single flag for everything to follow eventually, rather than adding per feature flags?

Of course things that need more granularity could define more flags in the future so it wouldn't need to be mutually exclusive to have a common flag and per feature flags.

Reply:
Yes I think starting with a single flag is the right approach. As you say it can be extended in the future.

We do need to emit a warning for everything that is unsupported but exported anyway so that the user knows what is happening.

Reply:
Added; please see: https://github.com/Autodesk/maya-usd/pull/952#discussion_r535728053